### PR TITLE
Reload: check stack slot and register class when deciding whether a move is a no-op

### PR DESCRIPTION
**Pull request by xclerc**
I think there might be a bug in [Reloadgen](https://github.com/ocaml-flambda/flambda-backend/blob/main/backend/reloadgen.ml#L68), which
is both quite unlikely and amd64-specific.

The issue is that `Reloadgen` accepts to leave both
operands of a move on the stack if they share the
same slot because it knows no instruction will be
emitted later, which means the constraints on the
operands are meaningless.

However, slots are bare integers, and each register
class indexes its slots from `0`. So, if we are unlucky
enough, it looks like a move with operands _(i)_ in
different register classes and _(ii)_ the same slot
will be considered as a nop when they should not.

(I think the same code pattern in arm64's [Reload](https://github.com/ocaml-flambda/flambda-backend/blob/main/backend/arm64/reload.ml)
is fine because on this architecture the operands
of a move are guaranteed to be in the same
register class.)

### CODE REVIEW

**Comment by xclerc** [comment]
I have fixes amd64's [Emit](https://github.com/ocaml-flambda/flambda-backend/blob/main/backend/amd64/emit.mlp#L510) in a similar way,
but in this module it may make more sense
to match on the `typ` fields rather than rely
on the register classes.

File: backend/amd64/emit.mlp [hunk 1]
```diff
@@ -507,7 +507,7 @@ let emit_global_label s =
   _label lbl
 
 let move (src : Reg.t) (dst : Reg.t) =
-  if src.loc <> dst.loc then
+  if src.loc <> dst.loc || Proc.register_class src <> Proc.register_class dst then
```
Comment (gretay-js) [inline backend/amd64/emit.mlp:510]: Why do we need this extra condition? I thought it should now be guaranteed from `reload`.

**Comment by stedolan** [comment]
I don't quite understand - aren't stack-to-stack moves no-ops as long as the slot is the same, even if the register classes differ? Or is this about reg-to-reg moves?

**Comment by xclerc** [comment]
I think the problem is that `Reg.(Stack (Local x))`
is not unique: there is one per register class.

**Review by gretay-js — COMMENTED** [review]
I think the upstream compiler maintains the invariant that the source and the destination of a `Imove` are of the same `machtype`, i.e. `arg.(0).typ` = `res.(0).typ`.  This implies that they have the same `register_class`. This is no longer the case in `flambda-backend` for amd64, after `float_of_bits` and `bits_of_float` intrinsics added in https://github.com/ocaml-flambda/flambda-backend/pull/296, which we implemented as an `Imove`.  It has already caused trouble in #385.  I think we should restore the invariant, and re-implement `caml_int64_bits_of_float_unboxed` and `caml_int64_float_of_bit_unboxed` using `Ispecific` instead.

The change proposed in this PR is correct and will prevent future mistakes. I am in favor of merging it, independently of the above-mentioned change for intrinsics. 

What happens in a similar situation with IRC?

**Comment by xclerc** [comment]
I think IRC itself is safe, but [that line](https://github.com/ocaml-flambda/flambda-backend/blob/main/backend/cfg/cfg.ml#L459) looks suspicious.

File: backend/amd64/emit.mlp (hunk 1 above)
Comment (xclerc) [reply backend/amd64/emit.mlp:510]: I don't think this extra condition is necessary indeed.

**Comment by xclerc** [comment]
(And we could do something better for `Move` [here](https://github.com/ocaml-flambda/flambda-backend/blob/main/backend/amd64/cfg_stack_operands.ml#L195).
I am adding a note to https://github.com/ocaml-flambda/flambda-backend/issues/758.)

**Comment by xclerc** [comment]
@gretay-js is arm64 correct? there is [apparently](https://github.com/ocaml-flambda/flambda-backend/blob/main/backend/arm64/emit.mlp#L674)
no code for moves between different types.

**Comment by gretay-js** [comment]
> is arm64 correct? there is [apparently](https://github.com/ocaml-flambda/flambda-backend/blob/main/backend/arm64/emit.mlp#L674)
> no code for moves between different types.

it's fine,  there is `fmov` for floats in https://github.com/ocaml-flambda/flambda-backend/blob/4e7dd0d2f70c3d95638a135d252977cd5baca829/backend/arm64/emit.mlp#L678

**Review by gretay-js — APPROVED** [review]

**Comment by xclerc** [comment]
And `fmov` accepts both operands as floats,
or one float and an integer, right?

**Comment by gretay-js** [comment]
> And `fmov` accepts both operands as floats, or one float and an integer, right?

apparently, it's for transfers between general purpose and FP registers. I don't think this part has changed with the bits_of_float intrinsics, which are amd64 only, so I think we should leave arm64 as is.

**Comment by stedolan** [comment]
Thanks for the explanation @xclerc and @gretay-js, I understand the issue now.

If the plan is to restore the invariant that `Imove` only happens between registers of the same machtype (which makes sense to me), then wouldn't we be better off with an assertion rather than a check here? (Or is the plan to have this patch as a temporary fix until the intrinsics are updated?)

**Comment by gretay-js** [comment]
> If the plan is to restore the invariant that `Imove` only happens between registers of the same machtype (which makes sense to me), then wouldn't we be better off with an assertion rather than a check here? (Or is the plan to have this patch as a temporary fix until the intrinsics are updated?)

yes, I'd prefer to merge it as a temporary fix, and then think about the intrinsics separately.

**Pull request merged**